### DISCUSSION TITLE
ddgr: changes to w3m settings are not persisted

### DISCUSSION THREAD

**Discussion by Rosika2** [netblue30/firejail#5420] [Q&A]
Hi all,  :wave:

I have a question regarding the terminal browser **w3m** triggered by **ddgr** (https://github.com/jarun/ddgr). All running in **firejail**.

> ddgr is a cmdline utility to search DuckDuckGo (html version) from the terminal

The command I use is:

```
cd /media/rosika/f14a27c2-0b49-4607-94ea-2e56bbf76fe1/DATEN-PARTITION/Dokumente/work/ddgr-2-0/ddgr-2.0_new_user-agent/
env BROWSER=w3m firejail --private=(pwd) ./ddgr
```

Well, that one functions well except for one little problem:

_example_ (running the above mentioned command):

**I type "bbc" as search word in "ddgr" and get the results:**

```
[...]
1.  BBC - Homepage [www.bbc.com]
        Breaking news, sport, TV, radio and a whole lot more. The BBC informs, educates and entertains - wherever you are, whatever your ageZ
 2.  BBC - Home [www.bbc.co.uk]
 [...]
```

 **I type "1"(for the first entry), hit "enter"** and then **w3m** opens up displaying the respective page. All good so far.

 - Upon quitting w3m however it shows: "Can't save history".
 - Also: _no bookmarks_ can be set and _no browser history_ is set either in **w3m**.
 - When trying to change settings in **w3m** it says: "There´s no config file ... config not saved". (But there is a config file.) :thinking:

Well, there is clearly an issue with **firejail** **permissions** as the following command would work:

```
cd /media/rosika/f14a27c2-0b49-4607-94ea-2e56bbf76fe1/DATEN-PARTITION/Dokumente/work/ddgr-2-0/ddgr-2.0_new_user-agent/
and env BROWSER=w3m firejail --noprofile --private=(pwd) ./ddgr
```

Of course I don´t want to run the processes using the "--noprofile" option. That was just for reference purposes.

Any idea which parameters/settings I have to change in w3m.profile to make it work in such a way that the settings can be changed and stored permanently :question:

Thanks a lot in advance.

Many greetings.
Rosika :slightly_smiling_face:

P.S.:

my system: Linux/Lubuntu 20.04.5 LTS, 64 bit

**Comment by kmk3** [comment]
@Rosika2 [on Oct 13](https://github.com/netblue30/firejail/discussions/5420#discussion-4470513):

> > ddgr is a cmdline utility to search DuckDuckGo (html version) from the terminal
> 
> The command I use is:
> 
> ```
> cd /media/rosika/f14a27c2-0b49-4607-94ea-2e56bbf76fe1/DATEN-PARTITION/Dokumente/work/ddgr-2-0/ddgr-2.0_new_user-agent/
> env BROWSER=w3m firejail --private=(pwd) ./ddgr
> ```
> 
> Well, that one functions well except for one little problem:
> 
> [...]
> 
> Well, there is clearly an issue with **firejail** **permissions** as the following command would work:
> 
> ```
> cd /media/rosika/f14a27c2-0b49-4607-94ea-2e56bbf76fe1/DATEN-PARTITION/Dokumente/work/ddgr-2-0/ddgr-2.0_new_user-agent/
> and env BROWSER=w3m firejail --noprofile --private=(pwd) ./ddgr
> ```

The program you're running is `ddgr`, so it is likely using ddgr.profile, which
redirects to googler-common.profile.

The latter seems to block more things than w3m.profile does, so that might be
the issue.

Does it work if you add `--private-bin=ddgr,env,python,python3 --profile=w3m`?

Does it work by adding the following changes to googler-common.local?

```firejail
ignore seccomp.block-secondary

# Allow perl (blacklisted by disable-interpreters.inc)
include allow-perl.inc

whitelist /usr/share/w3m

private-bin perl
```

Relates to #4333.

Cc: @pirate486743186 @rusty-snake (from #4333)

**Comment by Rosika2** [comment]
@kmk3:

Hi Kelvin, :wave:

Thanks for your very quick reply and for the explanation.

> Does it work if you add --private-bin=ddgr,env,python,python3 --profile=w3m?
I modified my command thus:

`cd /media/rosika/f14a27c2-0b49-4607-94ea-2e56bbf76fe1/DATEN-PARTITION/Dokumente/work/ddgr-2-0/ddgr-2.0_new_user-agent/; and env BROWSER=w3m firejail --private-bin=ddgr,env,python,python3 --profile=w3m --private=(pwd) ./ddgr`

... but now I get:

`Error: no suitable ./ddgr executable found` 

Hmm, :thinking: why might that be?

> Does it work by adding the following changes to googler-common.local?

Actually I don´t seem to have that one:

```
 ll /etc/firejail | grep googler
-rw-r--r-- 1 root root 1,3K Feb  9  2022 googler-common.profile
-rw-r--r-- 1 root root  313 Jul 11  2021 googler.profile

```
Thanks a lot anyway.

Many greetings from Rosika :slightly_smiling_face:

**Reply by kmk3** [reply]
@Rosika2 [on Oct 13](https://github.com/netblue30/firejail/discussions/5420#discussioncomment-3870454):

> @kmk3:
> 
> Hi Kelvin, wave
> 
> Thanks for your very quick reply and for the explanation.

No problem.

> > Does it work if you add --private-bin=ddgr,env,python,python3 --profile=w3m?
> > I modified my command thus:
> 
> `cd /media/rosika/f14a27c2-0b49-4607-94ea-2e56bbf76fe1/DATEN-PARTITION/Dokumente/work/ddgr-2-0/ddgr-2.0_new_user-agent/; and env BROWSER=w3m firejail --private-bin=ddgr,env,python,python3 --profile=w3m --private=(pwd) ./ddgr`
> 
> ... but now I get:
> 
> `Error: no suitable ./ddgr executable found`
> 
> Hmm, thinking why might that be?

This will probably always happen if using `private-bin` and running something
that is not in `$PATH` (which the current/working directory usually isn't).

Does it work with `--ignore=private-bin --profile=w3m` instead?

> > Does it work by adding the following changes to googler-common.local?
> 
> Actually I don´t seem to have that one:
> 
> ```
>  ll /etc/firejail | grep googler
> -rw-r--r-- 1 root root 1,3K Feb  9  2022 googler-common.profile
> -rw-r--r-- 1 root root  313 Jul 11  2021 googler.profile
> ```

I meant adding it to ~/.config/firejail/google-common.local (which may or may
not currently exist).

Note: This directory (~/.config/firejail) is the recommended place to put
user-specific modifications to firejail profiles, as it is part of the include
path used by the firejail `include` command.

**Comment by Rosika2** [comment]
@kmk3:  

Thanks again Kelvin. :slightly_smiling_face: 

`Does it work with --ignore=private-bin --profile=w3m instead?`

Alas, no:
The new command  yields the same result:
```

 cd /media/rosika/f14a27c2-0b49-4607-94ea-2e56bbf76fe1/DATEN-PARTITION/Dokumente/work/ddgr-2-0/ddgr-2.0_new_user-agent/; and env BROWSER=w3m firejail  --ignore=private-bin --profile=w3m instead --private=(pwd) ./ddgr; and cd

Error: no suitable instead executable found
```

> I meant adding it to ~/.config/firejail/google-common.local (which may or may
> not currently exist).

Ah, I see. Sorry, it seems I misunderstood. :blush: 

Well, as the respective file didn´t exist I created `google-common.local` in `/home/rosika/.config/firejail/` and put in  the following content:
```

ignore seccomp.block-secondary

# Allow perl (blacklisted by disable-interpreters.inc)
include allow-perl.inc

whitelist /usr/share/w3m

private-bin perl
```
But it doesn´t work either. I still get "Error: no suitable instead executable found".

Nevertheless: thanks so much for your help.

Many greetings
Rosika :slightly_smiling_face:

**Comment by Rosika2** [comment]
Hi again, :wave:

I found the solution :smiley: :

In` ddgr.profile` I had to comment out line 15 thus:

`#include disable-programs.inc`

That´s it. This way `w3m` has access to the bookmarks, history and settings are saved.

Of course I´d be interested in the explanation.
Plus: did I do it right or is there a better way :question: 

Thanks a lot and many greetings
Rosika :slightly_smiling_face:

**Reply by ghost** [reply]
i think you made a mess....
It's supposed to already just work if installed in PATH.
install it normally with the pkg manager or cp ddgr to /usr/local/bin

then
delete googler-common.local
and just try `env BROWSER=w3m firejail ddgr`

firejail profiles assume stuff are installed in PATH, you'll get problems if you try to put the executable elsewhere.

**Reply by rusty-snake** [reply]
https://github.com/netblue30/firejail/blob/65c5480f2f087ec483c240669af53bc0ebdaa2e2/etc/inc/disable-programs.inc#L1132

**Comment by Rosika2** [comment]
Hi again and thanks for your new replies, :wave:

@pirate486743186:

I did what you suggested. I copied `ddgr` to `/usr/local/bin` and of course I don´t have to define the path to it like I used to. I also deleted ` googler-common.local`.  And it works...
Bookmarks, history... everything is there. :+1: 

BUT: the fact remains I have to to comment out line 15 in `ddgr.profile`:
`#include disable-programs.inc`

Otherwise `w3m` has no access to bookmarks, history and settings. :thinking: 
That´s the only "mess" I made. :blush:  But until now this is the only way I can get it to work like it´s supposed to.

@rusty-snake:

Thanks for your post.
Yet I´m not sure what you´d like me to do... :thinking: 
Sorry.

Many thanks to all of you and many greetings
Rosika :slightly_smiling_face:

**Reply by ghost** [reply]
rusty snake means that w3m folder is black listed by disable-programs.inc

But that doesn't matter, because it's unblacklisted/whitelisted in googler-common.profile
It should already work for w3m, it just needs the BROWSER variable to be set to w3m.
only w3m is open by default, other browsers will need their own extra stuff.

you messed up somewhere.
and you must mean you edited googler-common.profile, not ddgr.profile
maybe the error is related to this?

**Reply by Rosika2** [reply]
@pirate486743186:

Thanks for your reply.

> rusty snake means that w3m folder is black listed by disable-programs.inc

I see. Right. I looked it up. I got `blacklist ${HOME}/.w3m` here as well.

> because it's unblacklisted/whitelisted in googler-common.profile

Yes, I got      `   noblacklist ${HOME}/.w3m   ` and `      whitelist ${HOME}/.w3m  ` here, like you said.

> and you must mean you edited googler-common.profile, not ddgr.profile

No, actually I did 
`#include disable-programs.inc` in `ddgr.profile` like I said. But this way it works...

Many greetings
Rosika :slightly_smiling_face:
